server/src/main/java/io/crate/planner/operators/JoinPlanBuilder.java

```diff
@@ -320,6 +320,71 @@ static List<JoinPair> convertImplicitJoinConditionsToJoinPairs(List<JoinPair> ex
         return newJoinPairs;
     }
 
+    static ArrayList<JoinPair> extractJoinPairsFromJoinConditions(List<JoinPair> explicitJoinPairs,
```

Comment:
This method looks very complex without any comment. Can you please add comments so I could understand why we need such a high complexity?
I wonder quite a bit as the issue sounds similar to https://github.com/crate/crate/commit/bf6e15652586cac9a3d45bb379be19e2dd5dd91c while this time its about implicit conditions.

Reply:
Yep indeed it's similar logic, you have part of a join condition referencing another "previous in join order" table , which produces a new join pair, and not only fixes the issue but also optimizes joins, since new inner join pairs can be produced.

I forgot to add comments indeed, will do, thx!

Reply:
I'm a bit confused:

https://github.com/crate/crate/commit/bf6e15652586cac9a3d45bb379be19e2dd5dd91c was about not discarding additional implicit join conditions.

Does the new method do the same (try to figure out some extra conditions) but on `explicitJoinPairs`? 

`explicitJoinPairs` construction has been recently updated in https://github.com/crate/crate/commit/e9624a9dcd64c40e8fd26b8285e187394e289adf#diff-0a97cff9d30c51ed0ce8e62390e7a51fbad67db9711ea6b59c241038af477755R318

Now I recalled a [comment](https://github.com/crate/crate/pull/13742#discussion_r1123461666) and wondering if we could change `RelationAnalyzer.extractJoinPair` to return list of pairs. We already have `relevantRelationsInOrder` there and currently return only last 2, maybe we can generate all pairs with  2 nested loops over `relevantRelationsInOrder`?

Reply:
This is a good idea, I completely missed this before, thx!

@seut What do you think of this idea, given that just extracting them to where clause is problematic (semantically) for outer joins (see: my other comment: https://github.com/crate/crate/pull/14029#issuecomment-1527375159)